smarts/core/scenario.py

```diff
@@ -317,61 +304,36 @@ def _discover_social_agents_info(
+        items = []
         for missions_file_path in glob.glob(file_match):
             with open(missions_file_path, "rb") as missions_file:
-                count = 0
-                missions = pickle.load(missions_file)
-
-            for mission_and_actor in missions:
-                # Each pickle file will contain a list of actor/mission pairs. The pairs
-                # will most likely be generated in an M:N fashion
-                # (i.e. A1: M1, A1: M2, A2: M1, A2: M2). The desired behavior is to have
-                # a single pair per concrete Scenario (which would translate to
-                # "per episode" when swapping)
-                assert isinstance(
-                    mission_and_actor.actor, sstudio_types.SocialAgentActor
-                )
-
-                actor = mission_and_actor.actor
-                extracted_mission = Scenario._extract_mission(
-                    mission_and_actor.mission, road_map
-                )
-                namespace = os.path.basename(missions_file_path)
-                namespace = os.path.splitext(namespace)[0]
-
-                setdefault(agent_bucketer, count, []).append(
-                    (
-                        SocialAgent(
-                            id=SocialAgentId.new(actor.name, group=namespace),
-                            name=actor.name,
-                            is_boid=False,
-                            is_boid_keep_alive=False,
-                            agent_locator=actor.agent_locator,
-                            policy_kwargs=actor.policy_kwargs,
-                            initial_speed=actor.initial_speed,
-                        ),
-                        extracted_mission,
+                actors, missions = pickle.load(missions_file)
+            namespace = os.path.basename(missions_file_path)
+            namespace = os.path.splitext(namespace)[0]
+
+            social_agent_infos = [
+                (agent.id, agent)
+                for agent in (
+                    SocialAgent(
+                        id=SocialAgentId.new(actor.name, group=namespace),
+                        name=actor.name,
+                        is_boid=False,
+                        is_boid_keep_alive=False,
+                        agent_locator=actor.agent_locator,
+                        policy_kwargs=actor.policy_kwargs,
+                        initial_speed=actor.initial_speed,
                     )
+                    for actor in actors
                 )
-                count += 1
+            ]
+            social_agent_missions = [
+                Scenario._extract_mission(mission, road_map) for mission in missions
+            ]
 
-        social_agents_info = []
-        for l in agent_bucketer:
-            social_agents_info.append(
-                {agent.id: (agent, mission) for agent, mission in l}
-            )
+            items.append((social_agent_infos, social_agent_missions))
```

Comment:
The idea here is that for every element in `items` list, you take every element `agent` in `social_agent_infos` and match it to every `social_agent_missions` giving you a cross product between the two array right?. I asked this since the old return type ensured this but the new one doesn't. This also means the cross product is happening at other level since i don't see that change in this pr.

Reply:
No, after discussion with @sah-huawei we made the decision to make the agents and missions explicit for each generated group rather than working with permutations. This throws the configuration explicitly to the scenario designer.

Reply:
lol, now I'm second guessing this!  maybe it's worth having one more (quick) discussion about this today?